modules/ROOT/edits/execution-engine.adoc

```diff
@@ -0,0 +1,152 @@
+
+[source, xml, linenums]
+----
+<flow>
+  <sftp:read path="personsArray.json" /> <1>
+  <http:request path="/persons" method="POST" /> <2>
+  <set-variable variableName="firstEntryName" value="#[payload[0].name]" /> <3>
+  <ee:transform ... /> <4>
+  <logger message="#[vars.firstEntryName]" /> <5>
+</flow>
+----
+
+According to the Proactor pattern, Mule submits the tasks as follows:
+
+<1> The blocking operation(`<sftp:read>`) executes in the IO pool.
+<2> `<http:request>` is a non blocking operation. Mule performs the request in the current thread. When the flow receives the response, Mule switches to the `CPU_LIGHT` pool.
+<3> `<set-variable>` operations should be fairly quick and stay in `CPU_LIGHT`. There is no thread switch.
+<4> `<ee:transform>`, this is potentially a computational heavy transformation, so Mule switches to the `CPU_INTENSIVE` pool.
+<5> Logger stays on `CPU_INTENSIVE`. There is no thread switch. +
+[NOTE]
+Due to optimizations regarding latency, thread switches are omitted when an IO or `CPU_INTENSIVE` task is followed by a `CPU_LIGHT` one. The reasoning behind this optimization is that executing said `CPU_LIGHT` task is most likely cheaper than the thread switch.
+
+== Configuration
+
+Mule runtime engine automatically configures the thread pools at startup time, by applying formulas that consider available resources such as CPU and memory. These formulas can be modified by editing the `MULE_HOME/conf/schedulers-pools.conf` file. However, MuleSoft doesn’t recommend changing the thread pools' configuration.
+
+Note that the configuration is global and affects the entire Mule instance.
+
+=== Configuration at the Application Level
+
+You can define the configuration of each pool and use it in a specific application by adding the following to your application code:
+
+[source,xml,linenums]
+----
+<ee:scheduler-pools gracefulShutdownTimeout="15000">
+   <ee:cpu-light
+           poolSize="2"
+           queueSize="1024"/>
+   <ee:io
+           corePoolSize="1"
+           maxPoolSize="2"
+           queueSize="0"
+           keepAlive="30000"/>
+   <ee:cpu-intensive
+           poolSize="4"
+           queueSize="2048"/>
+</ee:scheduler-pools>
+----
+
+==== Considerations
+
+Applying pool configurations at the application level causes Mule to create a completely new set of thread pools for the Mule app. This configuration does not change the default settings configured in the `scheduler-conf.properties` file, which is particularly important for on-premises deployments because many Mule apps can be deployed to the same Mule instance.
+
+Mulesoft recommends running mule using the default settings. If you use custom thread pool configurations, MuleSoft recommends that you perform load and stress testing with all applications involved in real-life scenarios to validate any change in the threading configurations and to understand how the thread pools work in Mule 4.
```

Comment:
```suggestion
Because Mulesoft recommends running Mule using the default settings, if you decide to customize thread pool configurations, be sure to perform load and stress testing with all applications, using real-life scenarios, and consult with MuleSoft Support before moving to production using customized settings.
```